typing/ctype.ml

```diff
@@ -533,6 +533,25 @@ exception Non_closed of type_expr * bool
 let free_variables = ref []
 let really_closed = ref None
 
+(* [free_vars_rec] collects the variables of the input type
+   expression into the [free_variables] reference. It is used for
+   several different things in the type-checker, with the following
+   bells and whistles:
+   - If [really_closed] is Some typing environment, types in the environment
+     are expanded to check whether the apparently-free variable would vanish
+     during expansion.
+   - We collect both type variables and row variables, paired with a boolean
+     that is [true] if we have a row variable.
+   - We do not count "virtual" free variables -- free variables stored in
+     the abbreviation of an object type that has been expanded (we store
+     the abbreviations for use when displaying the type).
+
+   The functions [free_vars] and [free_variables] below pass
+   the typing environment as an optional [?env] parameter.
```

Comment:
may receive a typing environment to be used for expansion

Reply:
I changed `pass the typing environment` into `receive the typing environment` -- the description for what the environment does is above.

Reply:
The nuance I was trying to convey it that you don't have to give the typing environment, an if you give it it sets `really_closed`.
Giving it changes the semantics of the function, which is not very clear in your explanation.
The overloading of the API is not very nice.

Reply:
New version:

```
   The functions [free_vars] and [free_variables] below receive
   a typing environment as an optional [?env] parameter and
   set [really_closed] accordingly.
```